dask/dataframe/io/parquet/arrow.py

```diff
@@ -321,30 +384,47 @@ def read_partition(
             columns += index
         if isinstance(piece, str):
             # `piece` is a file-path string
-            piece = pq.ParquetDatasetPiece(
-                piece, open_file_func=partial(fs.open, mode="rb")
-            )
+            pieces = [
+                pq.ParquetDatasetPiece(
+                    piece, open_file_func=partial(fs.open, mode="rb")
+                )
+            ]
         else:
-            # `piece` contains (path, row_group, partition_keys)
-            piece = pq.ParquetDatasetPiece(
-                piece[0],
-                row_group=piece[1],
-                partition_keys=piece[2],
-                open_file_func=partial(fs.open, mode="rb"),
-            )
-        df = piece.read(
-            columns=columns,
-            partitions=partitions,
-            use_pandas_metadata=True,
-            use_threads=False,
-            **kwargs.get("read", {}),
-        ).to_pandas(categories=categories, use_threads=False, ignore_metadata=True)[
-            list(columns)
-        ]
+            # `piece` contains (path, row_group, partition_keys, row_group_size)
+            (path, row_groups, partition_keys, row_group_size) = piece
+            if not isinstance(row_groups, list):
+                row_groups = [row_groups]
+            pieces = [
+                pq.ParquetDatasetPiece(
+                    path,
+                    row_group=row_group,
+                    partition_keys=partition_keys,
+                    open_file_func=partial(fs.open, mode="rb"),
+                )
+                for row_group in row_groups
+            ]
+
+        dfs = []
+        for piece in pieces:
+            df = piece.read(
+                columns=columns,
+                partitions=partitions,
+                use_pandas_metadata=True,
+                use_threads=False,
+                **kwargs.get("read", {}),
+            ).to_pandas(categories=categories, use_threads=False, ignore_metadata=True)[
+                list(columns)
+            ]
+
+            if index:
+                df = df.set_index(index)
+
+            if len(pieces) == 1:
+                return df
+
+            dfs.append(df)
 
-        if index:
-            df = df.set_index(index)
-        return df
+        return pd.concat(dfs, axis=0)
```

Comment:
I think that ideally the engine just has to tell us how to read a single row group, and that the core code can call it many times as necessary.  My hope is that this code can be generic enough to be used across all of the engines in a fairly general purpose way.